ably/state.go

```diff
@@ -120,85 +120,51 @@ type msgCh struct {
 	ch  chan<- error
 }
 
-func (q pendingEmitter) Len() int {
-	return len(q.queue)
-}
-
-func (q pendingEmitter) Less(i, j int) bool {
-	return q.queue[i].msg.MsgSerial < q.queue[j].msg.MsgSerial
-}
-
-func (q pendingEmitter) Swap(i, j int) {
-	q.queue[i], q.queue[j] = q.queue[j], q.queue[i]
-}
-
-func (q pendingEmitter) Search(msg *protocolMessage) int {
-	return sort.Search(q.Len(), func(i int) bool { return q.queue[i].msg.MsgSerial >= msg.MsgSerial })
+// Dismiss lets go of the channels that are waiting for an error on this queue.
+// The queue can continue sending messages.
+func (q *pendingEmitter) Dismiss() []msgCh {
+	cx := make([]msgCh, len(q.queue))
+	copy(cx, q.queue)
+	q.queue = nil
+	return cx
 }
 
 func (q *pendingEmitter) Enqueue(msg *protocolMessage, ch chan<- error) {
-	switch i := q.Search(msg); {
-	case i == q.Len():
-		q.queue = append(q.queue, msgCh{msg, ch})
-	case q.queue[i].msg.MsgSerial == msg.MsgSerial:
-		q.log.Warnf("duplicated message serial: %d", msg.MsgSerial)
-	default:
-		q.queue = append(q.queue, msgCh{})
-		copy(q.queue[i+1:], q.queue[i:])
-		q.queue[i] = msgCh{msg, ch}
+	if len(q.queue) > 0 {
+		expected := q.queue[len(q.queue)-1].msg.MsgSerial + 1
+		if got := msg.MsgSerial; expected != got {
+			panic(fmt.Sprintf("protocol violation: expected next enqueued message to have msgSerial %d; got %d", expected, got))
+		}
 	}
+	q.queue = append(q.queue, msgCh{msg, ch})
 }
 
 func (q *pendingEmitter) Ack(msg *protocolMessage, errInfo *ErrorInfo) {
-	if q.Len() == 0 {
-		return
-	}
-	ack, nack := 0, 0
-	// Ensure range [serial,serial+count] fits inside q.
-	switch i := q.Search(msg); {
-	case i == q.Len():
-		nack = q.Len()
-	case q.queue[i].msg.MsgSerial == msg.MsgSerial:
-		nack = i
-		ack = min(i+msg.Count, q.Len())
-	default:
-		nack = i + 1
-		ack = min(i+1+msg.Count, q.Len())
+	// The msgSerial from the server may not be the same we're waiting. If the
+	// server skipped some messages, they get implicitly NACKed. If the server
+	// ACKed some messages again, we ignore those. In both cases, we just need
+	// to correct the number of messages that get ACKed by that difference.
+	serialShift := int(msg.MsgSerial - q.queue[0].msg.MsgSerial)
```

Comment:
I realized I can't just do this, because the msgSerial distance between consecutive messages isn't always 1. Non-MESSAGE (e. g. ATTACH, PRESENCE...) protocol messages also increase msgSerial.

Will fix.

Reply:
See 9896fce.

Reply:
@SimonWoolf clarified this isn't true; see https://github.com/ably/docs/pull/1115.